### DISCUSSION TITLE
Error Messages when installing homebrew (Ubuntu 20.04.1)

### DISCUSSION THREAD

**Discussion by ignaraph** [Homebrew/discussions#406] [Linux]
Hey everyone, i tried installing Homebrew today on my Ubuntu System.

I ran the following : `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` as provided on the Homebrew website.

Im getting the following error messages: 
`error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
`
`fatal: the remote end hung up unexpectedly
`
`fatal: early EOF`
`fatal: index-pack failed`
`Error: Failure while executing; `git clone https://github.com/Homebrew/linuxbrew-core /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core` exited with 128.
`
`Error: Failure while executing; `/home/linuxbrew/.linuxbrew/bin/brew tap homebrew/core` exited with 1.
Failed during: /home/linuxbrew/.linuxbrew/bin/brew update --force
`
I tried looking for answers but couldn't solve it still. 
Linux+Gnu Version : Ubuntu 20.04.1 LTS

**Comment by SMillerDev** [comment]
Something is blocking your connection to Github, there's nothing homebrew can do to fix this for you. 

You can try doing just the `curl` part of the command with an added `-v` parameter to see if it gives you more details.

**Reply by chobotnice007** [reply]
You can download that file from GitHub via jsDelivr. The link will be [https://cdn.jsdelivr.net/gh/Homebrew/install/HEAD/install.sh](url).
Command: `/bin/bash -c "$(curl -fsSL https://cdn.jsdelivr.net/gh/Homebrew/install@master/install.sh)"`. But look there [https://www.jsdelivr.com/?docs=gh](url) because I am not sure if I put correct link there. @SMillerDev @ignaraph

**Comment by ignaraph** [comment]
Used the -v option but it did not provide more details.

`/bin/bash -c "$(curl -vfsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

**Reply by carlocab** [reply]
Try

    curl -vOL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh

**Reply by ignaraph** [reply]
That command did not give me any errors but upon trying to install a package or trying to execute the main page install command i get the same error messages.

**Reply by carlocab** [reply]
Sorry, I'm not sure what you mean. You'll need to be explicit about what commands you ran exactly after the above command. Showing the exact output of each command would also be useful.

**Reply by ignaraph** [reply]
After running `-vOL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh`  i did not get any issues.

Then i ran: `brew --version` and go this as output:
`Homebrew 2.7.1`
`Homebrew/linuxbrew-core N/A`

So i tried installing a package:
`brew install hello`

And i got the same error messages as the ones posted on my first message of this discussion

**Reply by carlocab** [reply]
Ok, try
```
cd $(brew --repo)/Library/Taps/homebrew
git clone https://github.com/Homebrew/linuxbrew-core.git homebrew-core
```

**Reply by ignaraph** [reply]
I dont have the "Taps" Directory, what i did was:
`cd homebrew/Library/Homebrew`
`pwd`
`/home/ignacio/homebrew/Library/Homebrew`

Then i used:
`git clone https://github.com/Homebrew/linuxbrew-core.git`

Got these error messages:
`error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
`
`fatal: the remote end hung up unexpectedly`
`fatal: early EOF
`
`fatal: index-pack failed
`

**Reply by carlocab** [reply]
Sorry, I made a mistake with my previous comment. Edited it accordingly. However, the mistake isn't relevant to the error you're seeing. It looks like you have a problem with either your built-in installation of curl, or your connection. Not much I can do to help you fix those, unfortunately.

**Reply by SMillerDev** [reply]
It's not an issue in curl for sure. It can also be an issue in the git binary or configuration.

**Reply by ignaraph** [reply]
How do you suggest i proceed with this issue? I'm really new to all of this. 

Thank you very much for your help and time!

**Reply by gromgit** [reply]
First post the output of the following:
```
/home/linuxbrew/.linuxbrew/bin/brew config
/home/linuxbrew/.linuxbrew/bin/brew doctor
```
Then run this command:
```
HOMEBREW_NO_ENV_FILTERING=1 GIT_TRACE_CURL=/tmp/git.trace.curl /home/linuxbrew/.linuxbrew/bin/brew update-reset
```
Upload `/tmp/git.trace.curl` to a GitHub Gist or https://pastebin.com/ , then post the URL of the gist/paste you created.

**Reply by ignaraph** [reply]
Hey these are the outputs of the first 2 commands:
![Output1](https://user-images.githubusercontent.com/70608198/103438064-ad022800-4c26-11eb-8932-c9fd05e822de.png)

![Output2](https://user-images.githubusercontent.com/70608198/103438067-b25f7280-4c26-11eb-919b-0c2d39309a9f.png)

URL to the pastebin: https://pastebin.com/tBj0czi4

**Reply by gromgit** [reply]
> Upload `/tmp/git.trace.curl` to a GitHub Gist or https://pastebin.com/ , then post the URL of the gist/paste you created.

I meant "upload the _contents_ of `/tmp/git.trace.curl`". The easiest way is to run `xclip -sel clip < /tmp/git.trace.curl` to copy the contents to your clipboard, then paste it into your existing pastebin.

**Reply by ignaraph** [reply]
![image](https://user-images.githubusercontent.com/70608198/103439100-0e7bc400-4c32-11eb-941a-37c907ff3735.png)

Xclip command doesnt exist and using the first command `tmp/git.trace.curl` logs an error in the teriminal. I tried with "sudo" but still no output.

**Reply by carlocab** [reply]
cat /tmp/git.trace.curl

? You may need `sudo`.

**Reply by ignaraph** [reply]
Thank you, ive managed to do it:


Gist url: https://gist.github.com/ignaraph/3d613ff484b5ae0b4b855643dc44e1f1

**Comment by gromgit** [comment]
OK, now:
```
GIT_TRACE_CURL=/tmp/git.trace.curl.core git clone https://github.com/Homebrew/linuxbrew-core /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core
```
then upload `/tmp/git.trace.curl.core` in the same way.

**Reply by ignaraph** [reply]
This is the output from `GIT_TRACE_CURL=/tmp/git.trace.curl.core git clone https://github.com/Homebrew/linuxbrew-core /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core`

![image](https://user-images.githubusercontent.com/70608198/103439638-9237af80-4c36-11eb-8a40-2cf84635c4c1.png)

The output from: `/tmp/git.trace.curl.core` Has 800k+ Lines. I cant manage to make a Gist out of it...

**Reply by ignaraph** [reply]
maybe set a specific depth?

**Reply by carlocab** [reply]
Try

    tar -zcvf trace.tar.gz /tmp/git.trace.curl.core

and then upload the `.tar.gz` file here. See the bottom of the input textbox for instructions. You can click on those instructions to reveal an upload dialogue.

**Reply by ignaraph** [reply]
The tar.gz file is 85mb, to upload it, it must be smaller than 10mb.

**Reply by gromgit** [reply]
Try `tail -1000 /tmp/git.trace.curl.core > /tmp/git.trace.curl.core.excerpt` to extract the last 1000 lines of the curl log, and upload `/tmp/git.trace.curl.core.excerpt`.

**Reply by ignaraph** [reply]
Done. 
Gist URL: https://gist.github.com/ignaraph/4a76fc268dc9a899ff4a6bfc6bbecd8e

**Reply by gromgit** [reply]
Hmmm. Your `git` can clearly talk to GitHub, enough that it can pull down a large part of the core tap, but it always fails the second fetch step for some TLS reason.

Are you using a web proxy on your system? What's the output of `env | grep -i proxy`?

**Comment by ignaraph** [comment]
The output was long so il put it in 2 images:

![Output4](https://user-images.githubusercontent.com/70608198/103441316-f95c6080-4c44-11eb-9548-688255556a31.png)


![output5](https://user-images.githubusercontent.com/70608198/103441319-fcefe780-4c44-11eb-91b0-b3c0a08eed1d.png)

**Reply by gromgit** [reply]
There's no `v` in the command I posted, which was to discover any environment variables with `proxy` in their names. Adding the `v` inverts the sense of the `grep`, hence you just dumped everything _except_ what we're looking for.

**Reply by ignaraph** [reply]
Sorry, i added the "v" option because i received no output from the `env | grep -i proxy` command and i thought you would want to see something. 

![image](https://user-images.githubusercontent.com/70608198/103441614-1bef7900-4c47-11eb-8b60-4c093b9699c3.png)

**Reply by gromgit** [reply]
Getting nothing is itself significant: That means your `git` operations aren't being "intercepted" by a proxy you configured.

That said, I'm out of ideas at this point, and I can't replicate your issue on my own 20.04 box. Since it's GnuTLS that is crapping out, one possibility is to manually install the Homebrew `git` (which uses OpenSSL instead) and force `brew` to use it to complete the initial update, but someone else will have to advise on that.

**Reply by carlocab** [reply]
I think in theory it should be possible to do this without having to tap linuxbrew-core by downloading the raw formula file from https://github.com/Homebrew/linuxbrew-core.

However, you would need to do this for all of `git`'s dependencies, starting from the root of `git`'s dependency graph. @gromgit, can you do `brew deps --tree git` on your 20.04 box? I tried following them manually but that's a bit unmanageable.

**Reply by gromgit** [reply]
Actually, now that you mention it, it may be as simple as preloading the core tap with the raw master tarball:
```
curl -sL https://github.com/Homebrew/linuxbrew-core/archive/master.tar.gz | tar xzf - -C /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core --strip-components=1
```
then:
```
HOMEBREW_NO_AUTO_UPDATE=1 brew install git
HOMEBREW_FORCE_BREWED_GIT=1 brew update-reset
```

**Reply by ignaraph** [reply]
It says there is no such file or directory:

![image](https://user-images.githubusercontent.com/70608198/103443410-f4a0a800-4c56-11eb-9c3f-a39e98403724.png)

**Reply by gromgit** [reply]
Huh, that means Git doesn't even get around to creating the repo directory.

New plan: Do a manual shallow clone to get the base stuff in, then unshallow it. Hope that gets past the GnuTLS hangup.
```
git clone --depth=1 https://github.com/Homebrew/linuxbrew-core /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core
git -C /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core fetch --unshallow
```
If that finishes without any errors, run the installer script one last time to finish off the installation.

_Sidenote:_ Please don't post screenshots, as they're pretty hard to read and impossible to copy-n-paste from. Post output as plain text.

**Reply by ignaraph** [reply]
Hey,

After running this command `git clone --depth=1 https://github.com/Homebrew/linuxbrew-core /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core` i received NO errors.

However when i ran this one `git -C /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core fetch --unshallow`
I got the following ouput:
error: RPC failed; curl 56 GnuTLS recv error (-110): The TLS connection was non-properly terminated.
fatal: the remote end hung up unexpectedly
fatal: protocol error: bad pack header

**Reply by gromgit** [reply]
OK, now that you have a working (though shallow) repo, try switching to Homebrew git to finish the unshallowing:
```
HOMEBREW_NO_AUTO_UPDATE=1 /home/linuxbrew/.linuxbrew/bin/brew install git
/home/linuxbrew/.linuxbrew/bin/git -C /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core fetch --unshallow
```
If all that works without errors, run the installer script again, then you should be able to use Homebrew normally from that point.

**Reply by ignaraph** [reply]
Hey, first command ran with no problems.

Second command didn't, this was the output:

error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: protocol error: bad pack header

**Reply by gromgit** [reply]
OK, now that both GnuTLS- and OpenSSL-enabled `git` clients are failing, it's no longer a software issue. There's something between your box and GitHub that's interfering with large Git transfers. You'll have to find someone closer to you to help track it down, because until you do, Homebrew will likely just keep failing at odd moments.

**Reply by TerrellAW** [reply]
Why does Homebrew need the entire commit history? Shouldn't the install script --depth=1 by default?

**Reply by p-linnane** [reply]
You're commenting in a thread from almost 5 years ago. We now have a JSON API, and don't require cloning Homebrew/core.

**Reply by p-linnane** [reply]
Since your GitHub profile says you're a student, let me offer you an additional learning opportunity. It's something called [open source etiquette](https://developer.mozilla.org/en-US/docs/MDN/Community/Open_source_etiquette).

I hope you'll read it, take it to heart, and reassess how you interact with maintainers of open source projects in the future. Those maintainers know far more about how and why the project is run the way it is. Making absurd, uninformed statements like "Git is not an installer guys, this is why it doesn't work" is not appropriate.

You're also now banned from interacting with Homebrew on GitHub.

**Comment by ignaraph** [comment]
## Solved!

I updated Ubuntu yesterday and was later able to install homebrew without any issues, thank you everyone for your help!

**Comment by jamieaa64** [comment]
For anyone else who gets here, I spent a good 3 hours on this. 

It turns out Virgin Media had a child filter on that was filtering out all of github.com because its possible to download "hacking software". Wierdly the connection would time out after 47% not straight away. Turning off this filter fixed it.

**Reply by AVaunt-consulting** [reply]
This was it for me as well! Thank you so much.

**Comment by MayHappybb** [comment]
Probably due to the slow network connection to github, try manually download the homebrew repository and then continuing using the bash script to finish the installation. Specifically, first use:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

when raised the error you mentioned, use ``ctrl+c`` to stop, and this script will create some path, go to the directory

```bash
cd /home/linuxbrew/.linuxbrew    # officially recommend prefix
```

then clone the repository manually

```bash
git clone https://github.com/Homebrew/brew Homebrew    # H in capital
```

this could be very slow, which is the reason caused that installation error, so maybe you could try clone this in another computer and ``scp`` it to here, when done, continue use the installation script:

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

and finished.
